builder/openstack/step_source_image_info.go

```diff
@@ -0,0 +1,100 @@
+package openstack
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/gophercloud/gophercloud/openstack/imageservice/v2/images"
+	"github.com/gophercloud/gophercloud/pagination"
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+)
+
+type StepSourceImageInfo struct {
+	SourceImage     string
+	SourceImageName string
+	ImageFilters    ImageFilterOptions
+}
+
+type ImageFilterOptions struct {
+	Filters    map[string]interface{} `mapstructure:"filters"`
+	MostRecent bool                   `mapstructure:"most_recent"`
+}
+
+func (s *StepSourceImageInfo) Run(_ context.Context, state multistep.StateBag) multistep.StepAction {
+	config := state.Get("config").(Config)
+	ui := state.Get("ui").(packer.Ui)
+
+	client, err := config.imageV2Client()
+
+	// if an ID is provided we skip the filter since that will return a single or no image
+	if s.SourceImage != "" {
+		state.Put("source_image", s.SourceImage)
+		return multistep.ActionContinue
+	}
+
+	params := &images.ListOpts{}
+
+	// build ListOpts from filters
+	if len(s.ImageFilters.Filters) > 0 {
+		errs := buildImageFilters(s.ImageFilters.Filters, params)
+		if len(errs.Errors) > 0 {
+			err := fmt.Errorf("Errors encountered in filter parsing.\n%s" + errs.Error())
+			state.Put("error", err)
+			ui.Error(errs.Error())
+			return multistep.ActionHalt
+		}
+	}
```

Comment:
This whole section that validates that the filter is correct should be in `Prepare` in `image_config.go` so any errors are surfaced during `packer validate`.

Reply:
Ahh yes this is correct. I will work on that refactor.

Reply:
My current approach: Build `ListOpts` within `Prepare` and save it to the `StageBag`. The `Run` phase of the `step_source_image_info` will check and use that if it exists in order to get the `source_image` from the client. Does this sound like a good approach to you?